html/user/openssl_sign_action.php

```diff
@@ -0,0 +1,148 @@
+<?php
+// This file is part of BOINC.
+// http://boinc.berkeley.edu
+// Copyright (C) 2008 University of California
+//
+// BOINC is free software; you can redistribute it and/or modify it
+// under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation,
+// either version 3 of the License, or (at your option) any later version.
+//
+// BOINC is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+// See the GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
+
+require_once("../inc/boinc_db.inc");
+require_once("../inc/user.inc");
+require_once("../inc/util.inc");
+//require_once('../inc/recaptchalib.php');
+
+// Needs to be defined, unknown use.
+check_get_args(array("tnow", "ttok"));
+
+// Check the user is online
+$user = get_logged_in_user();
+check_tokens($user->authenticator);
+
+// Optional: Recaptcha support
+//if ($recaptcha_private_key) {
+//    if (!boinc_recaptcha_isValidated($recaptcha_private_key)) {
+//        error_page(
+//            tra("Your reCAPTCHA response was not correct. Please try again.")
+//        );
+//    }
+//}
+
+// Input is passed in from the openssl_sign_form
+// sanitize_tags removes html elements from the user input
+// $user_data = htmlspecialchars((post_str("user_data", true)),ENT_XML1);
+
+// user data input without htmlspecialchars nor sanitize_tags
+$user_data = post_str("user_data", true);
+
+//Has the user entered data with length > 0 & <= 256?
+if ((strlen($user_data) > 0) && (strlen($user_data) <= 256)) {
+
+	// Retrieve the logged in user's UserID.
+	$userid = $user->id;
+	//$cpid = $user->cross_project_id;
+
+	// Create the message which will be signed.
+	$message_data = "$userid $user_data";
+
+	// Could add extra info into the message however this may cause issues for external systems.
+	// $timestamp = time();
+	// $message_data = "$userid $user_data $timestamp $cpid";
+
+	// How to generate required keys in /project/keys/ folder:
+	// openssl genpkey -algorithm RSA -out external_system_key_private.pem -pkeyopt rsa_keygen_bits:2048
+	// openssl rsa -pubout -in external_system_key_private.pem -out external_system_key_public.pem
+
+	// FAO project admins: Change the following 2 lines to the location of the keys you have personally generated.
+	// If the following keys do not exist, then the users will face (handled) error messages.
+	$private_key_path = "file:///root/project/keys/external_system_key_private.pem";
```

Comment:
This should use the existing key directory specified in `config.xml`. The same place the public code signing key is.

Reply:
Here is how to get the path out of `config.xml`:
```
$config = get_config();
$keydir = parse_config($config, "<key_dir>");
```

Reply:
https://github.com/BOINC/boinc/pull/2134/files#diff-91d6b04155bcf59e677280a695167c2fR51

Done.

Where is the ideal place for the public to get access to the openssl public key? Up to the project admin or is there an appropriate public download location for the pem file/text? Second copy/paste text area?